lib/src/modules/task.cc

```diff
@@ -543,5 +574,84 @@ void Task::processOutputAndUpdateMetadata(ActionResponse& response)
     }
 }
 
+static void defaultPurgeCallback(const std::string& dir_path)
+{
+    fs::remove_all(dir_path);
+}
+
+unsigned int Task::purge(
+    const std::string& ttl,
+    std::function<void(const std::string& dir_path)> purge_callback)
+{
+    unsigned int num_purged_dirs { 0 };
+    Timestamp ts { ttl };
+    if (purge_callback == nullptr)
+        purge_callback = &defaultPurgeCallback;
+
+    LOG_INFO("About to purge cached tasks from '{1}'; TTL = {2}",
+             task_cache_dir_, ttl);
+
+    lth_file::each_subdirectory(
+        task_cache_dir_,
+        [&](std::string const& s) -> bool {
+            fs::path dir_path { s };
+            LOG_TRACE("Inspecting '{1}' for purging", s);
+
+            boost::system::error_code ec;
+            pcp_util::lock_guard<pcp_util::mutex> the_lock { task_cache_dir_mutex_ };
+            auto last_update = fs::last_write_time(dir_path, ec);
+            if (ec) {
+                LOG_ERROR("Failed to remove '{1}': {2}", s, ec.message());
+            } else if (ts.isNewerThan(last_update)) {
+                LOG_TRACE("Removing '{1}'", s);
+
+                try {
+                    purge_callback(dir_path.string());
+                    num_purged_dirs++;
+                } catch (const std::exception& e) {
+                    LOG_ERROR("Failed to remove '{1}': {2}", s, e.what());
+                }
+            }
+
+            return true;
+        });
+
+    LOG_INFO(lth_loc::format_n(
+        // LOCALE: info
+        "Removed {1} directory from '{2}'",
+        "Removed {1} directories from '{2}'",
+        num_purged_dirs, num_purged_dirs, task_cache_dir_));
+    return num_purged_dirs;
+}
+
+//
+// Task cache directory purge task (private interface)
+//
+
+void Task::taskCacheDirPurgeTask()
```

Comment:
With some distance, I may look at pulling these tasks and threading into a separate class that both Task and RequestProcessor pull in.

An alternative is to have RequestProcessor have a separate list of things that require purging (such as the Task) class and run through that list as part of the purge task there.

Reply:
The purge functionality has a fair amount of duplication as well.